Initial Version of the Alica Test Utilities

StephanOpfer opened the pull request.

Commit:
- some first stubs for behaviour creator mocker

Commit:
- finished first version of TestBehaviourCreator

Commit:
added AlicaTestSupportUtility

Commit:
added class for alica test support utility

Commit:
changed comment, as not all behaviour creators are autogenerated

Commit:
added friend declaration towards AlicaTestSupportUtility
added dependency towards alica_test_support

Commit:
- support for mockup behaviours

Commit:
- implemented step mode for behaviours

Commit:
- refactored test library and improved API
- adapted alica tests
- minor improved engine

Commit:
update travis ci (melodic/beaver)

Commit:
- fix warning

Commit:
fix warnings

Commit:
fix another warning

Commit:
fix warning

Commit:
- fix warnings

veerajsk commented on alica_test_utility/include/alica/test/TestBehaviourCreator.h:
```diff
@@ -0,0 +1,56 @@
+    /**
+     * The default interface, used by the ALICA Engine to retrieve
+     * an instance of a behaviour. The TestBehaviourCreator stores created
+     * behaviours, so that there will only be one instance per ID.
```
Is this true now? since _behaviourCreateFunctions stores a function it should create a new instance everytime createBehaviour is called right?

veerajsk commented on alica_test_utility/include/alica/test/SuccessBehaviour.h:
```diff
@@ -0,0 +1,14 @@
+{
+class SuccessBehaviour : public IMockUpBehaviour
+{
+    explicit SuccessBehaviour(uint32_t iterationsBeforeSuccess = 0);
```
can take number of iterations as a template parameter. Then the test writer can set this template argument & still use it as a mock behaviour in the TestBehaviourCreator.

Commit:
- made fail and success behaviour usable via the TestBehaviourCreator

Commit:
adapted alica to the improved event_handling package

Commit:
improved debug output

Commit:
rename prepareStepping to makeBehaviourEventDriven

Commit:
make things more unique

Commit:
- fix dereferenceing unique pointer

abhi1287 commented on alica_engine/include/engine/BasicBehaviour.h:
```diff
@@ -16,8 +13,16 @@
 #include <vector>
 
+namespace essentials {
+    class ITrigger;
```
as a good design practice, a class from a different namespace must never be forward declared and rather the appropriate header should be pulled in.

StephanOpfer replied:
I tend to forward declare things nevertheless. Basically this sums up my arguments: https://gieseanw.wordpress.com/2018/02/25/the-joys-of-forward-declarations-results-from-the-real-world/
I did the same for our robotic soccer code base and also achieved a 15% faster build.

abhi1287 replied:
no one can deny the speed improvements, but my recommendation is from a design perspective. It's bad to forward declare things that are not in the same project/namespace. As things will not get recompiled when they are supposed to and throw up random errors at runtime. Have a look at the pros and cons here: https://google.github.io/styleguide/cppguide.html#Forward_Declarations

abhi1287 commented on alica_test_utility/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+		src/IdleBehaviour.cpp
+        src/IMockUpBehaviour.cpp
+		include/alica/test/SuccessBehaviour.hpp
+		include/alica/test/FailBehaviour.hpp
```
need formatting and don't need to add .hpp files.

StephanOpfer replied:
done

abhi1287 commented on alica_test_utility/CMakeLists.txt:
```diff
@@ -0,0 +1,34 @@
+        src/TestContext.cpp
+        src/BehaviourTrigger.cpp
+		src/IdleBehaviour.cpp
+        src/IMockUpBehaviour.cpp
```
"I" prefix generally stands for interface and not used to name implementation (cpp) files.

StephanOpfer replied:
Oh that was an Interface at the beginning. Did not rename it...

abhi1287 commented on alica_test_utility/include/alica/test/SuccessBehaviour.hpp:
```diff
@@ -0,0 +1,36 @@
+void SuccessBehaviour<iterationsBeforeSuccess>::run(void* msg)
+{
+    if (iterationsCounter() >= _iterationsBeforeSuccess && !this->isSuccess()) {
+        this->setSuccess();
```
"this" is almost always implied and can be skipped

StephanOpfer replied:
done

abhi1287 commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -0,0 +1,150 @@
+     * an inaccessible member variable of the AlicaContext, otherwise.
+     * @return Pointer to the AlicaEngine.
+     */
+    alica::AlicaEngine* getEngine();
```
We shouldn't give direct engine access to even unit tests of applications. Giving engine access means that our TestContext is not feature complete and application developers need to delve into how engine work to test applications. That's bad.

StephanOpfer replied:
No test is using the getEngine() thing of the TestContext, except for the already existing ones. I am definitely against having several friend classes declared in the engine code, just because the TestContext is not allowed to have a getEngine() interface. 
Though I can have a look, whether the existing test cases can use some more specific interface of the TestContext, but otherwise I would keep it.

StephanOpfer replied:
I adapted all tests to run without engine getter in TestContext.

abhi1287 commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -0,0 +1,150 @@
+     * @return True, if the given state was reached before the timeout. False, otherwise.
+     */
+    template <typename Rep, typename Period>
+    bool stepUntilStateReached(int64_t state, std::chrono::duration<Rep, Period> timeout);
```
consider the case: A single step of the engine can transition through multiple states.

StephanOpfer replied:
Well, if the engine jumps through multiple states, I expect this function to return false, because the agent did not stay in the requested state. Should I rename it to stepUntilStayInState?

abhi1287 replied:
Do we have a requirement for this? if yes, maybe we can overwrite MaxRuleApplications = 1 while in test mode and then this API should probably work.

StephanOpfer replied:
This method semantically originates from sootballs testcases. So either there is a requirement for this, or they were not aware of ALICA being able to skip several state at once and at the same time did not encounter any problem with that.

abhi1287 reviewed (commented):
Partial review till TestContext.h

Commit:
address PR comments:
- remove I from non-interface class
- remove engine getter (at first only temporarily to see, if we can live without it)
- cleanup CMakeList.txt

Commit:
- updated tests to run without engine getter
- introduced extra getDomainVariable method in TeamManager

StephanOpfer committed:
Merge branch 'json-plan-format' into sop_test_framework

Commit:
fix ci

Commit:
Merge branch 'sop_test_framework' of github.com:rapyuta-robotics/alica into sop_test_framework

Commit:
fix comporison of unsigned and signed int

abhi1287 commented on alica_engine/include/engine/AlicaContext.h:
```diff
@@ -29,7 +29,9 @@ namespace alica
 
```
stl headers should be coming in before project headers. maybe this got changed in json_plan_designer branch? 
the order should generally be:
- (<file>.h for the same <file>.cpp)
- C++ standard library headers
- dependent libraries' headers
- other project headers.

StephanOpfer replied:
Conclusion: we stick to the engines code guideline (system headers go last)

abhi1287 commented on alica_test_utility/include/alica/test/FailBehaviour.hpp:
```diff
@@ -0,0 +1,36 @@
+namespace alica::test
+{
+template <uint32_t iterationsBeforeFail>
+class FailBehaviour : public MockUpBehaviour
```
can merge this and success behaviour and do success or failure based on constructor argument,

abhi1287 commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -0,0 +1,235 @@
+
+    alica::PlanSelector* getPlanSelector();
+
+    alica::RunningPlan* makeRunningPlan(const alica::Plan* plan,const alica::Configuration* configuration);
```
why would alica application need to call these two methods?

abhi1287 commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -0,0 +1,235 @@
+     * Gives access to the root running plan node in the engine.
+     * @return Root Running Plan Node
+     */
+    const RunningPlan* getRootNode();
```
RunningPlan is an engine internal component.

abhi1287 commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -0,0 +1,235 @@
+    /** Gives access to the behaviour pool of the engine.
+     * @return BehaviourPool
+     */
+    const BehaviourPool& getBehaviourPool();
```
same question as before, this is internal too.

abhi1287 commented on alica_test_utility/include/alica/test/TestContext.h:
```diff
@@ -0,0 +1,235 @@
+    } else {
+        return "<ELEMENT-UNKNOWN>";
+    }
+    return element->getName();
```
missing null check on element

abhi1287 commented on alica_test_utility/src/TestContext.cpp:
```diff
@@ -0,0 +1,174 @@
+
+std::unique_ptr<BehaviourTrigger> TestContext::setBehaviourTrigger(int64_t behaviourID, int64_t configurationID)
+{
+    std::shared_ptr<alica::BasicBehaviour> behaviour = nullptr;
```
shall we put an assert that init shouldn't have been called already?

abhi1287 commented on alica_tests/include/test_alica.h:
```diff
@@ -25,17 +26,10 @@
-struct AlicaTestEngineGetter
-{
-    static alica::AlicaEngine* getEngine(alica::AlicaContext* ac) { return ac->_engine.get(); }
-};
```
That's a lot of change that we don't need. alica tests should be allowed to get to the engine and verify finer details. If you add everything to the TestContext it will become bloated and confuse application developers. TestContext is supposed to be used only by alica application tests and not alica tests yhemselves. You can revert all changes to the tests by keeping the engine getter a friend of AlicaContext as before, along with TestContext.

abhi1287 commented:
@StephanOpfer  what are you up to? Don't keep merging PR's without addressing the review comments. 
None of the tons of file changes in alica_tests was necessary.

StephanOpfer commented:
As mentioned on slack, I created a wrike task that tracks your comments: https://www.wrike.com/open.htm?id=561356566
I am addressing the comments currently on https://github.com/rapyuta-robotics/alica/tree/revise_alica_test_utility_api
Some of the changes happened in this merged PR actually improved the alica_tests, so not all were unnecessary.

Now I am trying to revise all tests and the test library API in a way that combines two "requirements":
1. not having the engine exposed in TestContext
2. not having access to the engine in alica_tests as well

The first requirement, I see as mandatory because you requested it. The second requirement is something that I try to achieve, but if it does not work, I will keep the extra EngineGetter-friend declaration in the engine. However, for the exposed BehaviourPool, for example, I already found a solution that makes all tests work without it and without an Engine-Getter-friend.
